server/server.go

```diff
@@ -187,8 +187,13 @@ type Server struct {
 	hotRegionStorage *storage.HotRegionStorage
 	// Store as map[string]*grpc.ClientConn
 	clientConns sync.Map
-	// Store as map[string]*tsopb.TSOClient
-	tsoClients sync.Map
+
+	tsoClientPool struct {
+		mux sync.Mutex
+		// Store as map[string]*tsopb.TSOClient
+		clients sync.Map
```

Comment:
+1

Reply:
This is because the most common operation is s.tsoClientPool.clients.Load(forwardedHost) which is outside of critical section and needs to be a access concurrently.

func (s *GrpcServer) getTSOForwardStream(ctx context.Context, forwardedHost string) (tsopb.TSO_TsoClient, error) {
	v, ok := s.tsoClientPool.clients.Load(forwardedHost)
	if ok {
		// This is the common case to return here
		return v.(tsopb.TSO_TsoClient), nil
	}

	s.tsoClientPool.mux.Lock()
	defer s.tsoClientPool.mux.Unlock()

	// Double check after entering the critical section
	v, ok = s.tsoClientPool.clients.Load(forwardedHost)
	if ok {
		return v.(tsopb.TSO_TsoClient), nil
	}

	// Now let's create the client connection and the forward stream
	client, err := s.getDelegateClient(ctx, forwardedHost)
	if err != nil {
		return nil, err
	@@ -2032,7 +2044,7 @@ func (s *GrpcServer) getTSOForwardStream(ctx context.Context, forwardedHost stri
		return nil, err
	}
	done <- struct{}{}
	s.tsoClientPool.clients.Store(forwardedHost, forwardStream)
	return forwardStream, nil
}

Reply:
The concurrent reads outside of critical section can be overlapped with the write inside of critical section.

Reply:
if we put v, ok := s.tsoClientPool.clients.Load(forwardedHost) into the critical section, it hurts the performance of most common operations, doesn't it?

Reply:
This pattern is common, if I'm right.

Reply:
You could:

```go
RLock()
TryLoad()
RUnlock()

Lock()
defer Unlock()
Do the rest of things.
```

Reply:
This is comparison between read lock and sync.Map.Load.

OK, I changed.